### ISSUE TITLE
match on code blocks causes syntax errors

### ISSUE THREAD

**Issue by p-avital** [rust-analyzer/rust-analyzer#1875]
Hello,
Found a false positive in ra-lsp's parsing of match patterns:
```rust
match { Some(()) } {
    Some(_) => {}
    None => {}
};
```
is valid rust, but will cause ra-lsp to misinterpret the matched value as the list of patterns, apparently.
Surrounding the block generating the value with parenthesis will cause ra-lsp to regain its composure.
```rust
match ({ Some(()) }) {
    Some(_) => {}
    None => {}
};
```
Have a nice day :)

**Comment by kjeremy** [comment]
The problem case seems to interpret the `{} Some(())` has a match arm:

```
MATCH_EXPR@[17; 35)
            MATCH_KW@[17; 22) "match"
            WHITESPACE@[22; 23) " "
            MATCH_ARM_LIST@[23; 35)
              L_CURLY@[23; 24) "{"
              WHITESPACE@[24; 25) " "
              MATCH_ARM@[25; 33)
                TUPLE_STRUCT_PAT@[25; 33)
                  PATH@[25; 29)
                    PATH_SEGMENT@[25; 29)
                      NAME_REF@[25; 29)
                        IDENT@[25; 29) "Some"
                  L_PAREN@[29; 30) "("
                  TUPLE_PAT@[30; 32)
                    L_PAREN@[30; 31) "("
                    R_PAREN@[31; 32) ")"
                  R_PAREN@[32; 33) ")"
              WHITESPACE@[33; 34) " "
              R_CURLY@[34; 35) "}"
```

I think the issue is that either  `expr_no_struct` is not doing the right thing or maybe we need to detect something else right before the call to `match_arm_list`.

**Comment by kjeremy** [comment]
@matklad 

I'm not sure if we typically go by the docs [here](https://doc.rust-lang.org/reference/expressions/match-expr.html) but they suggest that maybe the issue is in `expr_bp`?

**Comment by matklad** [comment]
I think it has to do with restriction handing. The best way is to pt a debug print to the `parse_match_expr` and see where it goes wrong

**Comment by kjeremy** [comment]
I don't think `expr_no_struct` is moving forward at https://github.com/rust-analyzer/rust-analyzer/blob/58dc3b16496cbcce14101605aa1214e263b91c87/crates/ra_parser/src/grammar/expressions/atom.rs#L361 so we're falling into the next check for `{` which starts `match_arm_list` but we're still looking at `{ Some() }`.

The call to `expr_bp` in `expr_no_struct` is returning `(None, NotBlock)` which is probably the issue (maybe something with `lhs`?).

**Comment by kjeremy** [comment]
The `None` being returned there is because we end up here: https://github.com/rust-analyzer/rust-analyzer/blob/58dc3b16496cbcce14101605aa1214e263b91c87/crates/ra_parser/src/grammar/expressions/atom.rs#L122

**Comment by kjeremy** [comment]
I think the fix may be to change that check to `if r.forbid_structs && la != IDENT`. That seems to pass existing tests and account for the new one.

**Comment by kjeremy** [comment]
Closed via #1950
